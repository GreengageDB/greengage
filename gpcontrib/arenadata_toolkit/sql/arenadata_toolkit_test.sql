----------------------------------------------------------------------------------------------------------
-- test helpers
----------------------------------------------------------------------------------------------------------
SET client_min_messages=WARNING;
-- start_matchsubs
-- m/(.*)prt_p\d{6}/
-- s/(.*)prt_p\d{6}/$1prt_pYYYYMM/
-- m/(.*)backup_\d{8}t\d{6}/
-- s/(.*)backup_\d{8}t\d{6}/$1backup_YYYYMMDDtHHMMSS/
-- end_matchsubs
-- function that mocks manual installation of arenadata_toolkit from bundle
CREATE FUNCTION mock_manual_installation() RETURNS VOID AS $$
BEGIN
	CREATE SCHEMA arenadata_toolkit;
	EXECUTE 'GRANT ALL ON SCHEMA arenadata_toolkit to ' || CURRENT_USER;
	EXECUTE FORMAT($fmt$CREATE TABLE arenadata_toolkit.db_files_history(collecttime TIMESTAMP WITHOUT TIME ZONE)
			WITH (appendonly=true, compresstype=zlib, compresslevel=9) DISTRIBUTED RANDOMLY
			PARTITION BY RANGE (collecttime)
			(
				PARTITION %1$I START (date %2$L) INCLUSIVE
				END (date %3$L) EXCLUSIVE
				EVERY (INTERVAL '1 month'),
				DEFAULT PARTITION default_part
			);$fmt$,
			'p' || to_char(NOW(), 'YYYYMM'),
			to_char(NOW(), 'YYYY-MM-01'),
			to_char(NOW() + interval '1 month','YYYY-MM-01'));
	CREATE TABLE arenadata_toolkit.daily_operation(field INT) WITH (appendonly=true) DISTRIBUTED RANDOMLY;
	CREATE TABLE arenadata_toolkit.operation_exclude(schema_name TEXT) WITH (appendonly=true) DISTRIBUTED RANDOMLY;
	CREATE EXTERNAL WEB TABLE arenadata_toolkit.db_files (field TEXT) EXECUTE 'echo 1' FORMAT 'TEXT';
END;
$$ LANGUAGE plpgsql;

-- view that returns information about tables that belongs to the arenadata_toolkit schema (and schema itself)
CREATE VIEW toolkit_objects_info AS
	SELECT oid as objid, 'schema'  as objtype, nspname as objname,
		'-'::"char" as objstorage,
		replace(nspacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_namespace WHERE nspname = 'arenadata_toolkit'
	UNION
	SELECT oid as objid, 'table' as objtype, relname as objname,
		relstorage as objstorage,
		replace(relacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_class WHERE relname IN (SELECT table_name FROM information_schema.tables WHERE table_schema = 'arenadata_toolkit')
	UNION
	SELECT oid as objid, 'proc' as objtype, proname as objname,
		'-'::"char" as objstorage,
		replace(proacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_proc WHERE pronamespace = (SELECT oid from pg_namespace WHERE nspname = 'arenadata_toolkit')
	;
----------------------------------------------------------------------------------------------------------

---------------------------------------------------------------------------------------------------------
-- test case 1: arenadata_toolkit was installed
----------------------------------------------------------------------------------------------------------

-- setup
SELECT mock_manual_installation();

-- check that created toolkit objects doesn't depend on extension
SELECT * FROM pg_depend d JOIN toolkit_objects_info objs ON d.objid = objs.objid AND d.deptype='e';

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;

CREATE EXTENSION arenadata_toolkit;
SELECT arenadata_toolkit.adb_create_tables();

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema after creating
-- extension and calling adb_create_tables
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;

-- check that toolkit objects now depends on extension
SELECT objname, objtype, extname, deptype FROM pg_depend d JOIN
	toolkit_objects_info objs ON d.objid = objs.objid JOIN
	pg_extension e ON d.refobjid = e.oid
WHERE d.deptype = 'e' AND e.extname = 'arenadata_toolkit' ORDER BY objname;

DROP EXTENSION arenadata_toolkit;
DROP SCHEMA arenadata_toolkit cascade;

----------------------------------------------------------------------------------------------------------
-- test case 2: arenadata_toolkit wasn't installed
----------------------------------------------------------------------------------------------------------

CREATE EXTENSION arenadata_toolkit;
SELECT arenadata_toolkit.adb_create_tables();

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema after creating
-- extension and calling adb_create_tables
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;

-- check that toolkit objects now depends on extension
SELECT objname, objtype, extname, deptype FROM pg_depend d JOIN
	toolkit_objects_info objs ON d.objid = objs.objid JOIN
	pg_extension e ON d.refobjid = e.oid
WHERE d.deptype = 'e' AND e.extname = 'arenadata_toolkit' ORDER BY objname;

DROP EXTENSION arenadata_toolkit;
DROP SCHEMA arenadata_toolkit cascade;

DROP FUNCTION mock_manual_installation();
DROP VIEW toolkit_objects_info;
RESET client_min_messages;
