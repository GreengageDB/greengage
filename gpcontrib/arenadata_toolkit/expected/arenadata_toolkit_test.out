----------------------------------------------------------------------------------------------------------
-- test helpers
----------------------------------------------------------------------------------------------------------
SET client_min_messages=WARNING;
-- start_matchsubs
-- m/(.*)prt_p\d{6}/
-- s/(.*)prt_p\d{6}/$1prt_pYYYYMM/
-- m/(.*)backup_\d{8}t\d{6}/
-- s/(.*)backup_\d{8}t\d{6}/$1backup_YYYYMMDDtHHMMSS/
-- end_matchsubs
-- function that mocks manual installation of arenadata_toolkit from bundle
CREATE FUNCTION mock_manual_installation() RETURNS VOID AS $$
BEGIN
	CREATE SCHEMA arenadata_toolkit;
	EXECUTE 'GRANT ALL ON SCHEMA arenadata_toolkit to ' || CURRENT_USER;
	EXECUTE FORMAT($fmt$CREATE TABLE arenadata_toolkit.db_files_history(collecttime TIMESTAMP WITHOUT TIME ZONE)
			WITH (appendonly=true, compresstype=zlib, compresslevel=9) DISTRIBUTED RANDOMLY
			PARTITION BY RANGE (collecttime)
			(
				PARTITION %1$I START (date %2$L) INCLUSIVE
				END (date %3$L) EXCLUSIVE
				EVERY (INTERVAL '1 month'),
				DEFAULT PARTITION default_part
			);$fmt$,
			'p' || to_char(NOW(), 'YYYYMM'),
			to_char(NOW(), 'YYYY-MM-01'),
			to_char(NOW() + interval '1 month','YYYY-MM-01'));
	CREATE TABLE arenadata_toolkit.daily_operation(field INT) WITH (appendonly=true) DISTRIBUTED RANDOMLY;
	CREATE TABLE arenadata_toolkit.operation_exclude(schema_name TEXT) WITH (appendonly=true) DISTRIBUTED RANDOMLY;
	CREATE EXTERNAL WEB TABLE arenadata_toolkit.db_files (field TEXT) EXECUTE 'echo 1' FORMAT 'TEXT';
END;
$$ LANGUAGE plpgsql;
-- view that returns information about tables that belongs to the arenadata_toolkit schema (and schema itself)
CREATE VIEW toolkit_objects_info AS
	SELECT oid as objid, 'schema'  as objtype, nspname as objname,
		'-'::"char" as objstorage,
		replace(nspacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_namespace WHERE nspname = 'arenadata_toolkit'
	UNION
	SELECT oid as objid, 'table' as objtype, relname as objname,
		relstorage as objstorage,
		replace(relacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_class WHERE relname IN (SELECT table_name FROM information_schema.tables WHERE table_schema = 'arenadata_toolkit')
	UNION
	SELECT oid as objid, 'proc' as objtype, proname as objname,
		'-'::"char" as objstorage,
		replace(proacl::text, CURRENT_USER, 'owner') AS objacl -- replace current_user to static string, to prevent test flakiness
		FROM pg_proc WHERE pronamespace = (SELECT oid from pg_namespace WHERE nspname = 'arenadata_toolkit')
	;
----------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------
-- test case 1: arenadata_toolkit was installed
----------------------------------------------------------------------------------------------------------
-- setup
SELECT mock_manual_installation();
 mock_manual_installation 
--------------------------
 
(1 row)

-- check that created toolkit objects doesn't depend on extension
SELECT * FROM pg_depend d JOIN toolkit_objects_info objs ON d.objid = objs.objid AND d.deptype='e';
 classid | objid | objsubid | refclassid | refobjid | refobjsubid | deptype | objid | objtype | objname | objstorage | objacl 
---------+-------+----------+------------+----------+-------------+---------+-------+---------+---------+------------+--------
(0 rows)

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;
               objname               | objtype | objstorage |      objacl      
-------------------------------------+---------+------------+------------------
 arenadata_toolkit                   | schema  | -          | {owner=UC/owner}
 daily_operation                     | table   | a          | 
 db_files                            | table   | x          | 
 db_files_history                    | table   | a          | 
 db_files_history_1_prt_default_part | table   | a          | 
 db_files_history_1_prt_p202306      | table   | a          | 
 operation_exclude                   | table   | a          | 
(7 rows)

CREATE EXTENSION arenadata_toolkit;
SELECT arenadata_toolkit.adb_create_tables();
 adb_create_tables 
-------------------
 
(1 row)

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema after creating
-- extension and calling adb_create_tables
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;
                          objname                           | objtype | objstorage |             objacl             
------------------------------------------------------------+---------+------------+--------------------------------
 __db_files_current                                         | table   | v          | {owner=arwdDxt/owner,=r/owner}
 __db_files_current_unmapped                                | table   | v          | {owner=arwdDxt/owner,=r/owner}
 __db_segment_files                                         | table   | v          | {owner=arwdDxt/owner,=r/owner}
 adb_collect_table_stats                                    | proc    | -          | {owner=X/owner}
 adb_create_tables                                          | proc    | -          | {owner=X/owner}
 adb_get_relfilenodes                                       | proc    | -          | {=X/owner,owner=X/owner}
 adb_hba_file_rules                                         | proc    | -          | {owner=X/owner}
 adb_hba_file_rules_view                                    | table   | v          | {owner=arwdDxt/owner}
 adb_relation_storage_size                                  | proc    | -          | {=X/owner,owner=X/owner}
 adb_relation_storage_size_on_segments                      | proc    | -          | {=X/owner,owner=X/owner}
 adb_skew_coefficients                                      | table   | v          | {owner=arwdDxt/owner,=r/owner}
 adb_vacuum_strategy                                        | proc    | -          | {owner=X/owner}
 adb_vacuum_strategy_newest_first                           | proc    | -          | {owner=X/owner}
 adb_vacuum_strategy_newest_last                            | proc    | -          | {owner=X/owner}
 arenadata_toolkit                                          | schema  | -          | {owner=UC/owner,=U/owner}
 daily_operation                                            | table   | a          | 
 db_files_current                                           | table   | h          | {owner=arwdDxt/owner,=r/owner}
 db_files_history                                           | table   | a          | {owner=arwdDxt/owner}
 db_files_history_1_prt_default_part                        | table   | a          | {owner=arwdDxt/owner}
 db_files_history_1_prt_pYYYYMM                             | table   | a          | {owner=arwdDxt/owner}
 db_files_history_backup_YYYYMMDDtHHMMSS                    | table   | a          | 
 db_files_history_backup_YYYYMMDDtHHMMSS_1_prt_default_part | table   | a          | 
 db_files_history_backup_YYYYMMDDtHHMMSS_1_prt_pYYYYMM      | table   | a          | 
 operation_exclude                                          | table   | a          | 
(24 rows)

-- check that toolkit objects now depends on extension
SELECT objname, objtype, extname, deptype FROM pg_depend d JOIN
	toolkit_objects_info objs ON d.objid = objs.objid JOIN
	pg_extension e ON d.refobjid = e.oid
WHERE d.deptype = 'e' AND e.extname = 'arenadata_toolkit' ORDER BY objname;
                objname                | objtype |      extname      | deptype 
---------------------------------------+---------+-------------------+---------
 __db_files_current                    | table   | arenadata_toolkit | e
 __db_files_current_unmapped           | table   | arenadata_toolkit | e
 __db_segment_files                    | table   | arenadata_toolkit | e
 adb_collect_table_stats               | proc    | arenadata_toolkit | e
 adb_create_tables                     | proc    | arenadata_toolkit | e
 adb_get_relfilenodes                  | proc    | arenadata_toolkit | e
 adb_hba_file_rules                    | proc    | arenadata_toolkit | e
 adb_hba_file_rules_view               | table   | arenadata_toolkit | e
 adb_relation_storage_size             | proc    | arenadata_toolkit | e
 adb_relation_storage_size_on_segments | proc    | arenadata_toolkit | e
 adb_skew_coefficients                 | table   | arenadata_toolkit | e
 adb_vacuum_strategy                   | proc    | arenadata_toolkit | e
 adb_vacuum_strategy_newest_first      | proc    | arenadata_toolkit | e
 adb_vacuum_strategy_newest_last       | proc    | arenadata_toolkit | e
(14 rows)

DROP EXTENSION arenadata_toolkit;
DROP SCHEMA arenadata_toolkit cascade;
----------------------------------------------------------------------------------------------------------
-- test case 2: arenadata_toolkit wasn't installed
----------------------------------------------------------------------------------------------------------
CREATE EXTENSION arenadata_toolkit;
SELECT arenadata_toolkit.adb_create_tables();
 adb_create_tables 
-------------------
 
(1 row)

-- show toolkit objects (and their grants) that belongs to arenadata_toolkit schema after creating
-- extension and calling adb_create_tables
SELECT objname, objtype, objstorage, objacl FROM toolkit_objects_info ORDER BY objname;
                objname                | objtype | objstorage |             objacl             
---------------------------------------+---------+------------+--------------------------------
 __db_files_current                    | table   | v          | {owner=arwdDxt/owner,=r/owner}
 __db_files_current_unmapped           | table   | v          | {owner=arwdDxt/owner,=r/owner}
 __db_segment_files                    | table   | v          | {owner=arwdDxt/owner,=r/owner}
 adb_collect_table_stats               | proc    | -          | {owner=X/owner}
 adb_create_tables                     | proc    | -          | {owner=X/owner}
 adb_get_relfilenodes                  | proc    | -          | {=X/owner,owner=X/owner}
 adb_hba_file_rules                    | proc    | -          | {owner=X/owner}
 adb_hba_file_rules_view               | table   | v          | {owner=arwdDxt/owner}
 adb_relation_storage_size             | proc    | -          | {=X/owner,owner=X/owner}
 adb_relation_storage_size_on_segments | proc    | -          | {=X/owner,owner=X/owner}
 adb_skew_coefficients                 | table   | v          | {owner=arwdDxt/owner,=r/owner}
 adb_vacuum_strategy                   | proc    | -          | {owner=X/owner}
 adb_vacuum_strategy_newest_first      | proc    | -          | {owner=X/owner}
 adb_vacuum_strategy_newest_last       | proc    | -          | {owner=X/owner}
 arenadata_toolkit                     | schema  | -          | {owner=UC/owner,=U/owner}
 daily_operation                       | table   | a          | {owner=arwdDxt/owner}
 db_files_current                      | table   | h          | {owner=arwdDxt/owner,=r/owner}
 db_files_history                      | table   | a          | {owner=arwdDxt/owner}
 db_files_history_1_prt_default_part   | table   | a          | {owner=arwdDxt/owner}
 db_files_history_1_prt_pYYYYMM        | table   | a          | {owner=arwdDxt/owner}
 operation_exclude                     | table   | a          | {owner=arwdDxt/owner}
(21 rows)

-- check that toolkit objects now depends on extension
SELECT objname, objtype, extname, deptype FROM pg_depend d JOIN
	toolkit_objects_info objs ON d.objid = objs.objid JOIN
	pg_extension e ON d.refobjid = e.oid
WHERE d.deptype = 'e' AND e.extname = 'arenadata_toolkit' ORDER BY objname;
                objname                | objtype |      extname      | deptype 
---------------------------------------+---------+-------------------+---------
 __db_files_current                    | table   | arenadata_toolkit | e
 __db_files_current_unmapped           | table   | arenadata_toolkit | e
 __db_segment_files                    | table   | arenadata_toolkit | e
 adb_collect_table_stats               | proc    | arenadata_toolkit | e
 adb_create_tables                     | proc    | arenadata_toolkit | e
 adb_get_relfilenodes                  | proc    | arenadata_toolkit | e
 adb_hba_file_rules                    | proc    | arenadata_toolkit | e
 adb_hba_file_rules_view               | table   | arenadata_toolkit | e
 adb_relation_storage_size             | proc    | arenadata_toolkit | e
 adb_relation_storage_size_on_segments | proc    | arenadata_toolkit | e
 adb_skew_coefficients                 | table   | arenadata_toolkit | e
 adb_vacuum_strategy                   | proc    | arenadata_toolkit | e
 adb_vacuum_strategy_newest_first      | proc    | arenadata_toolkit | e
 adb_vacuum_strategy_newest_last       | proc    | arenadata_toolkit | e
(14 rows)

DROP EXTENSION arenadata_toolkit;
DROP SCHEMA arenadata_toolkit cascade;
DROP FUNCTION mock_manual_installation();
DROP VIEW toolkit_objects_info;
RESET client_min_messages;
